qualtran/_infra/data_types.py

```diff
@@ -563,56 +580,109 @@ def is_symbolic(self) -> bool:
+    ) -> int:
+        """Returns the interpretation of the binary representation of `x` as an integer.
+
+        See class docstring section on "Classical Simulation" for more details on
+        the choice of this representation.
+
+        The returned value is an integer equal to `round(x * 2**self.num_frac)`.
+        That is, the input value `x` is converted to a fixed-point binary value
+        of `self.num_int` integral bits and `self.num_frac` fractional bits,
+        and then re-interpreted as an integer by dropping the decimal point.
+
+        For example, consider `QFxp(6, 4).to_fixed_width_int(1.5)`. As `1.5` is `0b01.1000`
+        in this representation, the returned value would be `0b011000` = 24.
+
+        For negative values, we use twos complement form. So in
+        `QFxp(6, 4, signed=True).to_fixed_width_int(-1.5)`, the input is `0b10.1000`,
+        which is interpreted as `0b101000` = -24.
+
+        Args:
+            x: input floating point value
+            require_exact: Raise `ValueError` if `x` cannot be exactly represented.
+            complement: Use twos-complement rather than sign-magnitude representation of negative values.
+        """
+        bits = self._fxp_to_bits(x, require_exact=require_exact, complement=complement)
         return self._int_qdtype.from_bits(bits)
 
+    def float_from_fixed_width_int(self, x: int) -> float:
+        """Helper to convert from the fixed-width-int representation to a true floating point value.
+
+        Here `x` is the internal value used by the classical simulator.
+        See `to_fixed_width_int` for conventions.
+
+        See class docstring section on "Classical Simulation" for more details on
+        the choice of this representation.
+        """
+        return x / 2**self.num_frac
+
     def __str__(self):
         if self.signed:
             return f'QFxp({self.bitsize}, {self.num_frac}, True)'
         else:
             return f'QFxp({self.bitsize}, {self.num_frac})'
 
-    # Experimental `fxpmath.Fxp` support.
-    # This support is currently experimental, and does not hook into the classical
-    # simulator protocol. Once the library choice for fixed-point classical real
-    # values is finalized, the code will be updated to use the new functionality
-    # instead of delegating to raw integer values (see above).
-
-    @property
     def fxp_dtype_template(self) -> Fxp:
         """A template of the `Fxp` data type for classical values.
 
-        -   op_sizing='same' and const_op_sizing='same' ensure that the returned object is not resized
-            to a bigger fixed point number when doing operations with other Fxp objects.
-        -   shifting='trunc' ensures that when shifting the Fxp integer to left / right; the digits are
-            truncated and no rounding occurs
-        -   overflow='wrap' ensures that when performing operations where result overflows, the overflowed
-            digits are simply discarded.
+        Usage:
```

Comment:
section formatting -- but you probably don't need this as a section. the docstring is always about usage :)

Reply:
removed